Recalculate the tax adjustments after customer address is updated

dangerdogz opened the pull request:
This is a fix for issue https://github.com/solidusio/solidus/issues/894

If you update an address in the order after the checkout (like by updating the Customer Details in the admin after an order is complete but before it's shipped), this can potentially put the Order in a new Tax zone and this is currently not supported.

Added by this pull request :

**Spree::Address**

A custom method to select only the fields related to Tax Zone matching (zipcode, country_id, state_id) to compare different addresses

**Spree::OrderContent**

Added a check in update_cart to recreate Tax adjustments in certain conditions

1) The order currently has no address (only an empty TaxLocation), we add one. Nothing to be done, it will be handled by the state transitions between :address and :delivery, and :delivery to :payment.

2) The order currently has an address, we update it but editing only trivial fields (firstname, lastname). We use address.tax_zone_attributes to compare the old and the new address and determine that nothing meaningful happened regarding tax zones, we do nothing.

3) The order currently has an address, we update the state / the country / the zipcode. We compute new Tax adjustments because we might be in a different zone.

4) old_tax_address will be  TaxLocation instead of an Address if there was no Address prior to the update, so we just return because it means that the order is either a VAT order and we don't need to recalculate Tax adjustments for now, or the order is a non-VAT order in the cart state, in which case we don't want to / we can't calculate Tax yet

**Potential code issues**

I wanted to use ActiveRecord::Dirty to determine if address was updated instead of using old_tax_zone / new_tax_zone but since Spree::Address factory if creating a new record each time we update a value, it's not possible to my knowledge. There might be a better way though

**To determine**

What if an order is paid? (like if you accept the pending payments, you double-check before shipping and notice that the address is wrong)

mamhoff commented on core/spec/models/spree/order_contents_spec.rb:
```diff
@@ -224,28 +224,34 @@
 
-    context "submits an address in a different tax zone than the current address" do
-      let(:shipping_address) { create :address, state_code: "NY" }
+    context "submits an address in a different tax zone than the current address", :tag => :focus do
```
That `focus` tag shouldn't go into the codebase :)

mamhoff commented on core/spec/models/spree/order_contents_spec.rb:
```diff
@@ -224,28 +224,34 @@
     end
 
-    context "submits an address in the same tax zone than the current address" do
+    context "submits an address in the same tax zone than the current address"do
```
Can you add that space back in?

mamhoff commented on core/spec/models/spree/order_contents_spec.rb:
```diff
@@ -224,28 +224,34 @@
 
       let(:params) do
-        { ship_address: updated_address}
+        { ship_address: updated_address, bill_address: updated_address}
```
And add a space here?

mamhoff commented on core/app/models/spree/address.rb:
```diff
@@ -73,11 +74,16 @@ def self.value_attributes(base_attributes, merge_attributes = nil)
 
+    def tax_zone_attributes
+      self.class.value_attributes(attributes.slice(*TAX_ZONE_ONLY_ATTRS))
+    end
```
Does this method have any unit tests?

dangerdogz replied:
I knew I forgot something, going to add some

mamhoff commented:
This is excellent work. I advised @dangerdogz IRL to do this at the model level, however I have an impression moving this to `Spree::Order` or `Spree::OrderUpdater` instead of `Spree::OrderContents` might be a worthwhile. 

There's one thing about the commits: While it's great to see your TDD, we want only commits that actually pass CI in the codebase. Can you group the commits that add functionality together with their new tests? Also, I agree with Hound most of the time.

Notwithstanding, good work!

mamhoff commented on core/app/models/spree/order_contents.rb:
```diff
@@ -57,6 +70,16 @@ def approve(user: nil, name: nil)
+      # Possible cases:
+      # [Old_address is a TaxLocation, New_address is a TaxLocation ]: non VAT order with no existing addresses who has not yet reached the address state, therefore we don't compute taxes
+      # [Old_address is a TaxLocation, New_address is an Adress ]: Order who reached at least the address state, therefore we can compute taxes
+      # [Old_address is an Adress, New_address is a TaxLocation ]: Should not happen, TaxLocation can be replaced by an Address but not the other way around. Order is a weird state and we do nothing because it should not happen
```
Can you re-format the comments so they fit in a GitHub code preview? :)

mamhoff commented on core/spec/models/spree/order_contents_spec.rb:
```diff
@@ -224,6 +224,72 @@
       }.to change { subject.order.total }
     end
 
+    context "given an order with existing addresses" do
```
These specs are great, and even if we decide to move the logic somewhere else, they should absolutely stay.

mamhoff commented on core/app/models/spree/address.rb:
```diff
@@ -15,6 +15,7 @@ class Address < Spree::Base
     alias_attribute :last_name, :lastname
 
     DB_ONLY_ATTRS = %w(id updated_at created_at)
+    TAX_ZONE_ONLY_ATTRS = %w(state_id country_id zipcode)
```
This is a naming thing: In my opinion, the `tax_zone` is a weird implementation-specific artifact. Can we name this constant `TAXATION_ATTRIBUTES` or something like that that does not imply `Spree::Zone` to be involved? Because it isn't (directly)

dangerdogz commented:
I fixed the various style issues, and addressed the feedback provided by @mamhoff 

Also added unit testing for the taxation_attributes method added

jhawthorn commented on core/app/models/spree/order_contents.rb:
```diff
@@ -57,6 +71,20 @@ def approve(user: nil, name: nil)
+      # Order is a weird state and we do nothing because it should not happen.
+      return if old_address.is_a?(Spree::Tax::TaxLocation) || new_address.is_a?(Spree::Tax::TaxLocation)
+
+      old_address.try(:taxation_attributes) != new_address.try(:taxation_attributes)
```
Sorry for the delay in reviewing. Could these be `.try!`?

Otherwise I think this is a great change and is a step in the right direction towards relying less on the state machine to get everything right.

dangerdogz replied:
Comments look wrong, so I need to fix that but otherwise : 

I don't think TaxLocation needs to implement taxation_attributes because we're not supposed to compare TaxLocations for the following reasons : if old_address is a TaxLocation it means that that it's either a VAT order and it's handled elsewhere or it's a standard order who hasn't reached the Address state so the taxes will be handled by the checkout

The only case where this is suppose to apply is when you have two addresses, so this would means that you're either updating the customer details from the backend, or that a registered user is updating his default addresses so the taxes showing in the cart would be updated.

Does it make mores sense? 

Good point about using try! though, we can be extra sure that we're comparing Spree::Addresses because it will crash otherwise

dangerdogz commented:
Reworded the comments a bit, did a double check on the tests, should be ready for review again once build is finished

jhawthorn commented:
Looking good to me :+1:

mamhoff commented:
Yep, this is great! 💯

mamhoff commented:
Thanks again Fréderic!
